Fixes #336 Add capability to specify proxy for Artifactory repository replications

Guent4 opened the pull request:
See https://github.com/jfrog/jfrog-client-go/pull/524 for the changes to the client to support the proxy field.

Things to note:
* This change would only support Artifactory 7.x+
* There is a pretty significant jump in the `jfrog-client-go` version so a bit worried about that.
* Even though all of the replication related tests passed for me, I was not able to get some of the other tests to pass.  Even when I tried running the tests on the `master` branch of the repo.  It might just be because of the Artifactory instance that I tested against.
* I wanted to added a test for this, but unfortunately was not able to figure out how to dynamically create a proxy via a REST API (couldn't find anything on https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API). If you can help point me in the right direction wrt this, I would love to add a test for configuring proxy. I have tested the changes using a GUI/manually created proxy.

chb0github commented on docs/resources/artifactory_push_replication.md:
```diff
@@ -58,6 +58,7 @@ The following arguments are supported:
     * `sync_properties` - (Optional)
     * `sync_statistics` - (Optional)
     * `path_prefix` - (Optional)
+    * `proxy` - (Optional)
```
Needs a description - preferrably something from official documentation @alexhung - where have you been getting yours?

alexhung replied:
I couldn't find any official description so I made this one up: https://github.com/jfrog/terraform-provider-artifactory/blob/master/docs/resources/artifactory_artifact_webhook.md?plain=1#L53

alexhung replied:
Also after having worked with proxy in couple of different scenarios now, I think we should call this attribute `proxy_key` which is more descriptive since this isn't the proxy hostname.

Guent4 replied:
I'll add a description (probably copying https://github.com/jfrog/terraform-provider-artifactory/blob/master/docs/resources/artifactory_artifact_webhook.md?plain=1#L53).

It's super confusing because the GET API for replications returns back the field named `proxy_ref` but the CREATE/UPDATE takes a field named `proxy`.

Since https://github.com/jfrog/terraform-provider-artifactory/blob/master/docs/resources/artifactory_artifact_webhook.md?plain=1#L53 I'm slightly in favor of keeping it as `proxy` for the sake of consistency.  If you guys feel strongly about renaming this to `proxy_ref`, I can do that

alexhung replied:
@Guent4 Let's keep this as `proxy` for consistency, as you mentioned.

chb0github commented on docs/resources/artifactory_replication_config.md:
```diff
@@ -58,6 +58,7 @@ The following arguments are supported:
     * `sync_properties` - (Optional)
     * `sync_statistics` - (Optional)
     * `path_prefix` - (Optional)
+    * `proxy` - (Optional)
```
Need a description

chb0github commented on docs/resources/artifactory_single_replication_config.md:
```diff
@@ -58,6 +58,7 @@ The following arguments are supported:
 * `sync_properties` - (Optional)
 * `sync_statistics` - (Optional)
 * `path_prefix` - (Optional)
+* `proxy` - (Optional)
```
description

chb0github commented on go.mod:
```diff
@@ -7,7 +7,7 @@ require (
 	github.com/hashicorp/terraform-json v0.13.0 // indirect
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.7.0
-	github.com/jfrog/jfrog-client-go v0.27.0
+	github.com/jfrog/jfrog-client-go v1.10.0
```
@alexhung - this needs to go.

alexhung replied:
Agree. I'll create a ticket and schedule it soon.

alexhung replied:
@Guent4 Some context with regards to `jfrog-client-go`. We have been moving away from using `jfrog-client-go` for a while now and use Resty client instead. Replication plus couple of other resources are still using structs from `jfrog-client-go`. Our plan is to replace them with local structs and remove all dependency of that package.

chb0github commented on pkg/artifactory/resource_artifactory_pull_replication.go:
```diff
@@ -27,9 +27,9 @@ func resourceArtifactoryPullReplication() *schema.Resource {
+func unpackPullReplication(s *schema.ResourceData) *utils.UpdateReplicationBody {
 	d := &ResourceData{s}
-	replicationConfig := new(utils.ReplicationBody)
+	replicationConfig := new(utils.UpdateReplicationBody)
```
So, not your fault, but this is mixing types from the JFROG client go and using some locals too. I would prefer if you literally C&P the `utils.*` into the file and ditch the dependency. I'll let @alexhung chime in

alexhung replied:
Agree, @chb0github. As first step of removing dependency on `jfrog-client-go`, we should do what you describe.

Guent4 replied:
Ok, sounds good.  I'll do as chb0github suggests and copy out the types into this repo and not add more dependencies on `jfrog-client-go`

chb0github commented on pkg/artifactory/resource_artifactory_push_replication.go:
```diff
@@ -95,6 +102,10 @@ var pushReplicationSchema = map[string]*schema.Schema{
 	},
+	"proxy": {
+		Type:     schema.TypeString,
+		Optional: true,
```
Description. Also, if I am not mistaken, if a blank string is sent it doesn't actually change anything @alexhung

alexhung replied:
Correct. If empty string (or `null` in JSON payload) is sent to Artifactory, it will be ignored. To reset `proxy` to nothing, you need to send an non-existent proxy key. Anything will do as long as it isn't empty string or `null`.

You can use this function to do what you need: https://github.com/jfrog/terraform-provider-artifactory/blob/master/pkg/artifactory/repositories.go#L847

chb0github reviewed (changes requested):
This is more like a draft. But it's a good start. Maybe a member of the team can help carry it across the finish line. 

However, we actually want to remove dependency on the jf client.

alexhung reviewed (changes requested):
Thanks for the PR. As @chb0github commented, we need the following:
- Not add additional dependency with `jfrog-go-client`
- Add `Description` to schema attribute
- Need checks in test cases

To add proxy to the system, check out this test case: https://github.com/jfrog/terraform-provider-artifactory/blob/master/pkg/artifactory/resource_artifactory_remote_repository_test.go#L648

I recommend moving those proxy setup/teardown code to `util_test.go` so they can be shared.

Guent4 commented:
Thank you guys for all the feedback!  I see that you have assigned people to get this over the finish line (thanks!); do you have a rough estimate on how long it would take to get something released (in the scale of days vs weeks vs months)?  I'm mainly just curious and want to give help give an estimate to my team at work.  I have built a local version of the provider and am using that for now so not a huge rush.

Please let me know if there's anything else you would like me contribute to here with this PR.  More than happy to help :)

alexhung commented:
@Guent4 Once you push the updates from the reviews, @chb0github, @maheshjfrog, and @dasmanas will review again. We also need an entry in `CHANGELOG.md` which one of us can add for you.

Time scale-wise, my guess is days (and week or 2) depending on the updates.

Guent4 commented:
Ok sounds good.  I'll take a stab at the updates on Monday.

Time scale-wise, that sounds perfect!

Guent4 commented:
Thanks again for advice on how to fix up the PR! I have updated the PR to not rely on the updated `jfrog-client-go` and added proxy to the existing test.

Couple notes:
* Can someone double check that I'm correctly using the `handleResetWithNonExistantValue` function correctly here: https://github.com/jfrog/terraform-provider-artifactory/pull/337/files#diff-24148f5bffa37b38b5952641902e1d354e80282d856899340143035eb96f00d5R209
* When running the tests, most of them failed when I used the 30 day trial license as instructed in the README.  I was able to get around it by stealing an unused Enterprise license that we weren't using.  Not sure if I was doing something wrong or not
* When running `make acceptance` to run the tests, the tests consistently timed out on my machine.  I had to add a `-timeout 30m` to the `go test`
* The tests would transiently fail on some `post-test destroy`.  For example:
```
=== RUN   TestAccAccessTokenFixedDateGood
    resource_artifactory_access_token_test.go:141: Step 1/1 error: Error running apply: exit status 1
        
        Error: end date must be in the future, but is 2022-02-28 15:30:18 -0500 EST
        
          with artifactory_access_token.foobar,
          on terraform_plugin_test.tf line 10, in resource "artifactory_access_token" "foobar":
          10: resource "artifactory_access_token" "foobar" {
        
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: err: Resource id[artifactory_access_token.foobar] not found
--- FAIL: TestAccAccessTokenFixedDateGood (12.83s)
```
Re-running would most likely see the test pass.

Guent4 commented:
I rebased my branch with the updated `master` branch and noticed that:
* The timeout issue I mentioned above was fixed
* However, it seems like the new changes in `master` are causing new failures in the tests.  For example:
```
=== RUN   TestAccVirtualRepositoryWithInvalidProjectKeyGH318
2022/02/28 19:48:58.531145 ERROR RESTY 
401 POST http://localhost:8082/access/api/v1/projects
{
  "errors" : [ {
    "code" : "UNAUTHORIZED",
    "message" : "Unsupported authentication method Basic"
  } ]
}, Attempt 1
    util_test.go:145: 
        401 POST http://localhost:8082/access/api/v1/projects
        {
          "errors" : [ {
            "code" : "UNAUTHORIZED",
            "message" : "Unsupported authentication method Basic"
          } ]
        }
--- FAIL: TestAccVirtualRepositoryWithInvalidProjectKeyGH318 (0.07s)
```

chb0github commented:
> I rebased my branch with the updated `master` branch and noticed that:
> 
> * The timeout issue I mentioned above was fixed
> * However, it seems like the new changes in `master` are causing new failures in the tests.  For example:
> 
> ```
> === RUN   TestAccVirtualRepositoryWithInvalidProjectKeyGH318
> 2022/02/28 19:48:58.531145 ERROR RESTY 
> 401 POST http://localhost:8082/access/api/v1/projects
> {
>   "errors" : [ {
>     "code" : "UNAUTHORIZED",
>     "message" : "Unsupported authentication method Basic"
>   } ]
> }, Attempt 1
>     util_test.go:145: 
>         401 POST http://localhost:8082/access/api/v1/projects
>         {
>           "errors" : [ {
>             "code" : "UNAUTHORIZED",
>             "message" : "Unsupported authentication method Basic"
>           } ]
>         }
> --- FAIL: TestAccVirtualRepositoryWithInvalidProjectKeyGH318 (0.07s)
> ```

Huh. How are you starting and testing? I noticed that the start script has been deleted out of the bin directory - that would typically use docker and a fixed version.  

Is it all the tests failing with 401? 

So, I know that the platform team except RT (Xray/Projects/etc.) Doesn't support basic and that they were moving away from basic auth. I am going to guess that, whatever version of RT you're using has officially removed support for basic auth. 

@maheshjfrog and @alexhung  and @dasmanas - If that's the case, all tests are broke until this is fixed. You can change the auth to API key or TOKEN, but Automating that as part of a test sweet will take some work 

@danielmkn implemented this for testing in our automated builds but not for local testing. @Guent4 - Please confirm it's all the tests

Guent4 commented on pkg/artifactory/resource_artifactory_push_replication.go:
```diff
@@ -2,22 +2,53 @@ package artifactory
 )
 
-type PushReplication struct {
+type replicationBody struct {
```
Ok sure I'll make the change to export it

Guent4 replied:
Fixed

Guent4 commented:
> Huh. How are you starting and testing? I noticed that the start script has been deleted out of the bin directory - that would typically use docker and a fixed version.
I first tried to just use the `run-artifactory.sh` script.  However I had trouble with both Artifactory instances spinning up.  It might be because I'm using WSL on Windows and Windows does funky stuff.  So I ended up doing the following:
1. Added an `artifactory.lic` file with a enterprise license
2. Commented out spinning up the second Artifactory instance in `docker-compose.yml`.  The version of Artifactory that is being spun up is `7.27.10`
3. Ran lines 1 through 16 in `run-artifactory.sh` to spin up the Artifactory instance
4. Exported the env vars needed to run the tests (`ARTIFACTORY_URL`, `ARTIFACTORY_USERNAME`, `ARTIFACTORY_PASSWORD`, `TF_ACC`)
5. Ran `make acceptance` to run all of the tests

> Is it all the tests failing with 401?
The following tests failed with the `Bearer authentication is required` 401.
```
TestAccFederatedRepoWithProjectAttributesGH318
TestAccFederatedRepositoryWithInvalidProjectKeyGH318
TestAccLocalGenericRepositoryWithProjectAttributesGH318
TestAccLocalGenericRepositoryWithInvalidProjectKeyGH318
TestAccLocalGenericRepositoryWithInvalidProjectEnvironmentsGH318
TestAccRemoteRepositoryWithProjectAttributesGH318
TestAccRemoteRepositoryWithInvalidProjectKeyGH318
TestAccVirtualGenericRepositoryWithProjectAttributesGH318
TestAccVirtualRepositoryWithInvalidProjectKeyGH318
```

chb0github commented:
Yeah, this PR is blocked until this auth issue is fixed @danielmkn - What's status on implementing Bearer token auth?

chb0github commented:
Wait a sec - It isn't that RT is rejecting basic - it's that the project testing part is - we have a minimal project client to test out that portion of the integration.  I don't know how this ever worked, unless projects recently made this change. 

I guess inevitably this still has to change or none of the project integration and be tested and I know basic auth is being sunset on all products

chb0github commented:
@Guent4 - rebase with master - those tests have been commented out while we sort the auth issue. Then your tests should pass

Guent4 commented:
@chb0github thank you for the update!  I have rebased with master and all the tests passed except one transient failure:
```
=== RUN   TestAccAccessTokenFixedDateGood
    resource_artifactory_access_token_test.go:141: Step 1/1 error: Error running apply: exit status 1
        
        Error: end date must be in the future, but is 2022-03-03 15:25:25 -0500 EST
        
          with artifactory_access_token.foobar,
          on terraform_plugin_test.tf line 10, in resource "artifactory_access_token" "foobar":
          10: resource "artifactory_access_token" "foobar" {
        
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: err: Resource id[artifactory_access_token.foobar] not found
--- FAIL: TestAccAccessTokenFixedDateGood (12.86s)
```

I think this PR is ready for re-review.  Thanks for all the help you guys have provided!

chb0github commented:
ok, @Guent4 - To get you unblocked, you can generate and use a Bearer token, found at `${host}/ui/admin/configuration/security/access_tokens` - which can only be generated from the UI, and use that for local testing. 

When you have that token, set an env var before running the tests locally:

`export ARTIFACTORY_ACCESS_TOKEN = eyJ2ZXIiOiIyIiwidHlwIjoiSldUIiwiYWxnIjoiUlMyNTYiLCJraWQiOiJVcjJXUDA0b3NjR2hEVnp5ejF4cTIxSlVpOVV3R05wbUdRWi1HN3NwUHo0In0.eyJleHQiOiJ7XCJyZXZvY2FibGVcIjpcInRydWVcIn0iLCJzdWIiOiJqZmFjQDAxZTNzeW1idGJiMjc4MXZlcTBnOWsxeXFmXC91c2Vyc1wvY2hyaXN0aWFuYiIsInNjcCI6ImFwcGxpZWQtcGVybWlzc2lvbnNcL2FkbWluIiwiYXVkIjpbImpmcnRAKiIsImpmYWNAKiIsImpmZXZ0QCoiL6f43972dqZmNvbkAqIl0sImlzcyI6ImpmZmVAMDAwIiwiZXhwIjoxNjc3ODg5MTUxLCJpYXQiOjE2NDYzNTMxNTEsImp0aSI6IjA5MjliYmY0LWRjMWYtNGJmNC1iMDkwLTJmODI5MTEyNDlhMiJ9.S2wsg_vDlBDaTvSWN6ozkbXIOEeSbGr72HHZ1D6jm0KKuOiKpswf_PX0x0FJbB-Z-C6VyAex28D0HloRoMmteAdFx8RWie90koX2ENhfOBgDJofJRsXpSLpSoN92OIJYcsG_jDjeDMG7VxdhPQ8QqP-umZi1hHWUkEQ5hJDIvhDjckoa3TL7h3dyc9q8p6ESnwpTb765aNpWvzfQyoXajyxrUWnEdl2eQ5eEeQMrwzh1r9UCoeqHpmzMipb_FeH133w4kU0m8z8QPoUw5zA5c4z5zbT7UXc3um5pTemoRIjqmenRNfu6Z-N_mYVsAhNuBRFMZ0Fjo8InPhUQ1NJf_Q`

Then run:

```
export TF_ACC=1
make acceptance
```

@danielmkn  will be updating the documentation to specify this, and our automated testing will not require it for everything. But, this will get you unblocked right away

Guent4 commented:
@chb0github Thanks for the update!  I tried exactly as you recommended by setting `ARTIFACTORY_ACCESS_TOKEN` and saw the same behavior.  The following test failed:
``
=== RUN   TestAccAccessTokenFixedDateGood
    resource_artifactory_access_token_test.go:141: Step 1/1 error: Error running apply: exit status 1
        
        Error: end date must be in the future, but is 2022-03-04 08:50:38 -0500 EST
        
          with artifactory_access_token.foobar,
          on terraform_plugin_test.tf line 10, in resource "artifactory_access_token" "foobar":
          10: resource "artifactory_access_token" "foobar" {
        
    testing_new.go:70: Error running post-test destroy, there may be dangling resources: err: Resource id[artifactory_access_token.foobar] not found
--- FAIL: TestAccAccessTokenFixedDateGood (13.14s)
```

From what I've seen before, the failure is transient.  Is that correct?

chb0github commented:
We run these tests with every build and we never see this.

How are you testing it? Are you using the startup script in /scripts?

Please give us step by step what you're doing to test and we will try to replicate here

Guent4 commented:
I followed the steps I listed in https://github.com/jfrog/terraform-provider-artifactory/pull/337#issuecomment-1055818604

danielmkn commented:
@Guent4 we are looking into it!

maheshjfrog commented:
@Guent4 ... Regarding the test failure message.
ie: "Error: end date must be in the future, but is 2022-03-04 08:50:38 -0500 EST"
My local tests doesn't exhibit this failure (ran multiple time to check for any transient behavior).
Test case ie: TestAccAccessTokenFixedDateGood ... adds 10 seconds to the current time for the end_date attribute value.

Would like to know if there is any timezone difference between the Artifactory and the testing instance. 
My understanding, you have local Artifactory so there is no timezone difference. Please let me know. Thanks.

Guent4 commented:
Maybe it's just slow on my computer (the run took longer than the expected 20 min (but less than 30 which I set the test timeout to)?  If the issue isn't happening for you, then it's probably my setup being inconsistent/unreliable/slow.

There is no timezone difference between the Artifactory and the testing instance.  I'm running the Artifactory container and the tests on the same machine (both in WSL on Windows).

maheshjfrog commented:
I see.
To valid the slowness, May be you can try increasing the timeout value from 10 seconds to higher value in the test case.
ie: [Edit fixedDateGood function](https://github.com/jfrog/terraform-provider-artifactory/blob/3ce83518b1f04e6c61ee48e08b04eb03b81cf059/pkg/artifactory/resource_artifactory_access_token_test.go#L123)

Guent4 commented:
I'll try Monday morning.  I have to set up a new environment because I lost access to the environment I was testing in.

Have a wonderful weekend :)

chb0github commented:
@maheshjfrog and @danielmkn - we can give him a test instance. Or, alternatively, please pull his code, test and run locally. We can't accept the code unless someone on jfrog attests that it's good

Guent4 commented:
FWIW, I re-ran the tests on a new setup.  All the tests passed, however it took 46 min

maheshjfrog commented:
Thanks for the update @Guent4 ...
I ran the `make acceptance` tests on my local machine with your code. ie: https://github.com/Guent4/terraform-provider-artifactory Branch: AddReplicationProxySupport
Took 5 mins and all tests passed.

Guent4 commented:
Thank you for verifying!  It must be a WSL thing since it’s essentially running a Ubuntu VM on the Windows laptop which probably impacts performance of the tests

chb0github commented:
Since we've manually verified, let's make sure these changes get merged in. Thank you @Guent4 for your contribution. Sorry for the delay

Guent4 commented:
It was not a delay at all!  Thank you guys for all the guidance :)

maheshjfrog commented:
I see the PR is blocked for merge because CHANGELOG not updated. @Guent4 can you update the CHANGELOG.md file... we can quickly merge. Thanks.

Guent4 commented:
@maheshjfrog would you guys prefer 2.21.1 or 2.22.0?  Technically new feature is being added, but can also view this as a bugfix for the lacking feature.

maheshjfrog commented:
Since this is a new feature ... it will be 2.22.0. Thanks @Guent4

Guent4 commented:
Thanks everyone!
